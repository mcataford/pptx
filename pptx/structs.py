import typing
from dataclasses import dataclass

from xml.dom.minidom import Element


@dataclass
class TextRun:
    """
    Representation of a text segment within a paragraph.

    See Paragraph documentation.
    """

    text: str
    props: typing.Dict[str, str]

    @classmethod
    def from_xml(cls, element: Element) -> "TextRun":
        text = element.getElementsByTagName("a:t")[0]
        props_xml = element.getElementsByTagName("a:rPr")
        props = props_xml[0].attributes if props_xml else None

        extracted_properties: typing.Dict[str, str] = {}

        if props:
            for prop_index in range(props.length):
                extracted_properties[props.item(prop_index).name] = props.item(
                    prop_index
                ).value

        extracted_text = text.firstChild.nodeValue if text.hasChildNodes() else ""

        return cls(text=extracted_text, props=extracted_properties)


@dataclass
class Paragraph:
    """
    Representation of a paragraph within the text body of a shape.

    See: https://web.archive.org/web/20220627001219/http://officeopenxml.com/drwSp-text-paragraph.php
    """

    runs: typing.List[TextRun]

    @classmethod
    def from_xml(cls, element: Element) -> "Paragraph":
        runs = [
            TextRun.from_xml(child) for child in element.getElementsByTagName("a:r")
        ]
        return cls(runs=runs)


@dataclass
class TextBody:
    """
    Representation of the text body of a shape.

    See: https://web.archive.org/web/20220627002331/http://officeopenxml.com/drwSp-text.php
    """

    paragraphs: typing.List[Paragraph]

    @classmethod
    def from_xml(cls, element: Element) -> "TextBody":
        paragraphs = [
            Paragraph.from_xml(child) for child in element.getElementsByTagName("a:p")
        ]
        return cls(paragraphs=paragraphs)


@dataclass
class Shape:
    """
    Representation of a single element on a slide.

    See: https://web.archive.org/web/20231226205058/http://officeopenxml.com/drwShape.php
    """

    id: str
    name: str
    is_text_box: bool
    text: TextBody

    @classmethod
    def from_xml(cls, element: Element) -> "Shape":
        shape_non_visual_props = element.getElementsByTagName("p:nvSpPr")[0]
        common_non_visual_props = shape_non_visual_props.getElementsByTagName(
            "p:cNvPr"
        )[0]
        common_non_visual_shape_properties = (
            shape_non_visual_props.getElementsByTagName("p:cNvSpPr")[0]
        )

        shape_name = common_non_visual_props.getAttribute("name")
        shape_id = common_non_visual_props.getAttribute("id")
        is_text_box = common_non_visual_shape_properties.getAttribute("txBox") == "1"

        text_nodes = element.getElementsByTagName("p:txBody")[0]

        text_body = TextBody.from_xml(text_nodes)

        return cls(
            id=shape_id, name=shape_name, is_text_box=is_text_box, text=text_body
        )


@dataclass
class Slide:
    """
    Representation of a single slide part of a slideset.
    """

    raw_xml: str
    shapes: typing.List[Shape]

    @classmethod
    def from_xml(cls, element: Element) -> "Slide":
        shapes: list[Shape] = []

        for shape_xml in element.getElementsByTagName("p:sp"):
            shapes.append(Shape.from_xml(shape_xml))

        return cls(raw_xml=element.toxml(), shapes=shapes)


@dataclass
class Presentation:
    """
    Representation of the unpacked content of the slideset.
    """

    slides: typing.List[Slide]
