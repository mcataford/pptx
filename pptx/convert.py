"""
Conversion utilities to handle PPTX files.
"""
import zipfile
import tempfile


import pathlib

from xml.dom.minidom import parse

from structs import Slide, Presentation


def presentation_from_file(path: str) -> Presentation:
    unpacked_path = tempfile.mkdtemp()

    with zipfile.ZipFile(path, "r") as z:
        z.extractall(unpacked_path)

    slides_dir = pathlib.Path(unpacked_path, "ppt", "slides")

    slides: list[Slide] = []

    for file in slides_dir.iterdir():
        if file.suffix != ".xml":
            # File is likely not a slide.
            continue

        slides.append(_slide_from_file(file))

    return Presentation(slides=slides)


def _slide_from_file(path: pathlib.Path) -> Slide:
    """ """
    parsed_slide = parse(str(path))
    return Slide.from_xml(parsed_slide)
