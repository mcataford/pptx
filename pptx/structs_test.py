import pathlib
from xml.dom.minidom import parse, Element
import zipfile

import pytest

from pptx.structs import Slide


@pytest.fixture()
def get_sample(tmp_path):
    """Retrieves a sample slideset from the /samples directory by name."""

    def _get_sample_slide(label: str, index: int) -> Element:
        sample_path = pathlib.Path("samples", f"{label}.pptx")
        with zipfile.ZipFile(sample_path, "r") as z:
            z.extractall(tmp_path)

        slide_path = pathlib.Path(tmp_path, "ppt", "slides", f"slide{index+1}.xml")

        return parse(str(slide_path))

    return _get_sample_slide


def test_extracts_no_shapes_from_empty_slide(get_sample):
    """Shapes store content, an empty slide has no content nor shapes."""
    sample = get_sample("single_empty_slide", 0)

    slide = Slide.from_xml(sample)

    assert len(slide.shapes) == 0


def test_extracts_shape_from_nonempty_slide(get_sample):
    """A slide with exactly one element has exactly one shape."""
    sample = get_sample("text_samples", 0)

    slide = Slide.from_xml(sample)

    assert len(slide.shapes) == 1


def test_extracts_multiple_shapes_from_slide(get_sample):
    sample = get_sample("text_samples", 1)

    slide = Slide.from_xml(sample)

    assert len(slide.shapes) == 3


def test_extracts_paragraphs_from_shape_with_text(get_sample):
    """A slide containing a text box with one block of text has one paragraph."""
    sample = get_sample("text_samples", 0)

    slide = Slide.from_xml(sample)
    shape = slide.shapes[0]
    shape_text = shape.text

    assert len(shape_text.paragraphs) == 1


def test_extracts_text_runs_from_shape_with_text(get_sample, snapshot):
    """A paragraph is made of runs."""
    sample = get_sample("text_samples", 0)

    slide = Slide.from_xml(sample)
    shape = slide.shapes[0]
    shape_text = shape.text
    paragraphs = shape_text.paragraphs
    runs = paragraphs[0].runs

    assert len(runs) == 1

    run = runs[0]

    assert run.text == snapshot
    assert run.props == snapshot


@pytest.mark.parametrize(
    "run_text,expected_props",
    [["boldness", {"b": "1"}], ["italics", {"i": "1"}]],
    ids=["bold", "italics"],
)
def test_extracts_text_formatting_properties_from_shape_with_text(
    get_sample, snapshot, run_text, expected_props
):
    sample = get_sample("text_samples", 2)

    paragraph = Slide.from_xml(sample).shapes[0].text.paragraphs[0]
    runs = paragraph.runs

    selected_run = [run for run in runs if run.text == run_text][0]

    for prop in expected_props:
        assert expected_props[prop] == selected_run.props[prop]
