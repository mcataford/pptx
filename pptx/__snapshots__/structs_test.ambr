# serializer version: 1
# name: test_extracts_text_runs_from_shape_with_text
  'Sample text'
# ---
# name: test_extracts_text_runs_from_shape_with_text.1
  dict({
    'lang': 'en-CA',
    'sz': '1800',
  })
# ---
